cvmfs/catalog_mgr_rw.h

```diff
@@ -160,8 +174,72 @@ class WritableCatalogManager : public SimpleCatalogManager {
    * Directories don't have extended attributes at this point.
    */
   XattrList empty_xattrs;
+
+  /**
+   * Balancer thresholds
+   */
+  const bool is_balanced_;
+  const unsigned max_weight_;
+  const unsigned min_weight_;
+  const unsigned balance_weight_;
 };  // class WritableCatalogManager
 
+
+template <class CatalogMgrT>
+class CatalogBalancer {
+ public:
+  typedef typename CatalogMgrT::catalog_t catalog_t;
+  explicit CatalogBalancer(CatalogMgrT *catalog_mgr)
+    : catalog_mgr_(catalog_mgr) { }
+
+  void Balance(catalog_t *catalog = NULL);
+
+
+ protected:
+  struct VirtualNode {
+    vector<VirtualNode> children;
+    unsigned weight;
+    DirectoryEntry dirent;
+    string path;
+    bool is_new_nested_catalog;
+
+    void ExtractChildren(CatalogMgrT *catalog_mgr);
+    void CaltulateWeight();
+    VirtualNode(const string &path, CatalogMgrT *catalog_mgr)
+      : children(), weight(1), dirent(), path(path),
+        is_new_nested_catalog(false) {
+      catalog_mgr->LookupPath(path, kLookupSole, &dirent);
+    }
+    VirtualNode(const string &path, const DirectoryEntry &dirent,
+                CatalogMgrT *catalog_mgr)
+      : children(), weight(1), dirent(dirent), path(path),
+        is_new_nested_catalog(false) {
+      if (!IsCatalog() && IsDirectory())
+        ExtractChildren(catalog_mgr);
+    }
+    bool IsDirectory() { return dirent.IsDirectory(); }
+    bool IsCatalog() { return is_new_nested_catalog ||
+        dirent.IsNestedCatalogMountpoint(); }
+  };
+  typedef typename CatalogBalancer<CatalogMgrT>::VirtualNode virtual_node_t;
+
+ private:
+  void OptimalPartition(VirtualNode *virtual_node);
+  void AddCvmfsCatalogFile(string path);
+  DirectoryEntryBase CreateEmptyContentDirectoryEntryBase(string name,
```

Comment:
That's a very long name (but precise).  Please think 5-10 minutes if you can find a shorter name.  If nothing comes up, we'll leave it as is.